lib/hsa/mcwamp_hsa.cpp

```diff
@@ -271,6 +271,143 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
 };
 
 
+namespace
+{
+    template<typename P>
+    inline
+    ELFIO::section* find_section_if(ELFIO::elfio& reader, P p)
+    {
+        using namespace std;
+
+        const auto it = find_if(
+            reader.sections.begin(), reader.sections.end(), move(p));
+
+        return it != reader.sections.end() ? *it : nullptr;
+    }
+
+    inline
+    std::vector<std::string> copy_names_of_undefined_symbols(
+        const ELFIO::symbol_section_accessor& section)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        vector<string> r;
+
+        for (auto i = 0u; i != section.get_symbols_num(); ++i) {
+            // TODO: this is boyscout code, caching the temporaries
+            //       may be of worth.
+            string name;
+            Elf64_Addr value = 0;
+            Elf_Xword size = 0;
+            Elf_Half sect_idx = 0;
+            uint8_t bind = 0;
+            uint8_t type = 0;
+            uint8_t other = 0;
+
+            section.get_symbol(
+                i, name, value, size, bind, type, sect_idx, other);
+
+            if (sect_idx == SHN_UNDEF && !name.empty()) {
+                r.push_back(std::move(name));
+            }
+        }
+
+        return r;
+    }
+
+    inline
+    std::pair<ELFIO::Elf64_Addr, ELFIO::Elf_Xword> find_symbol_address(
+        const ELFIO::symbol_section_accessor& section,
+        const std::string& symbol_name)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        static constexpr pair<Elf64_Addr, Elf_Xword> r{0, 0};
+
+        for (auto i = 0u; i != section.get_symbols_num(); ++i) {
+            // TODO: this is boyscout code, caching the temporaries
+            //       may be of worth.
+            string name;
+            Elf64_Addr value = 0;
+            Elf_Xword size = 0;
+            Elf_Half sect_idx = 0;
+            uint8_t bind = 0;
+            uint8_t type = 0;
+            uint8_t other = 0;
+
+            section.get_symbol(
+                i, name, value, size, bind, type, sect_idx, other);
+
+            if (name == symbol_name) return make_pair(value, size);
+        }
+
+        return r;
+    }
+
+    inline
+    void associate_code_object_symbols_with_host_allocation(
+        const ELFIO::elfio& reader,
+        const ELFIO::elfio& self_reader,
+        ELFIO::section* code_object_dynsym,
+        ELFIO::section* process_symtab,
+        hsa_agent_t agent,
+        hsa_executable_t executable)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        if (!code_object_dynsym || !process_symtab) return;
+
+        const auto undefined_symbols = copy_names_of_undefined_symbols(
+            symbol_section_accessor{reader, code_object_dynsym});
+
+        for (auto&& x : undefined_symbols) {
```

Comment:
We could, but this would only be beneficial if we had multiple code objects per agent (otherwise having multiple unmerged symbols with the same name in the same code object / TU would be an ODR violation / not possible), which I do not believe we do at this point in time. The underlying assumption here is that the mapping from code-object undefined symbol to loader allocated host symbol is bijective. However, I may be missing something about how executables is managed at the moment...

Reply:
If an application is linked to a shared library with GPU code then it's possible that an agent will have more than one code objects (i.e. one from the application's objects and another from the shared library).